src/coreclr/src/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -2,158 +2,536 @@
+        /// <exception cref="NullReferenceException">The address of location is a null pointer.</exception>
+        public static long Increment(ref long location) =>
+            Add(ref location, 1);
 
-        public static long Decrement(ref long location)
-        {
-            return Add(ref location, -1);
-        }
+        /// <summary>Increments a specified variable and stores the result, as an atomic operation.</summary>
+        /// <param name="location">The variable whose value is to be incremented.</param>
+        /// <returns>The incremented value.</returns>
+        /// <exception cref="NullReferenceException">The address of location is a null pointer.</exception>
+        [CLSCompliant(false)]
+        public static ulong Increment(ref ulong location) =>
+            Add(ref location, 1);
+        #endregion
 
-        /// <summary>
-        /// Implemented: int, long, float, double, Object, IntPtr
-        /// </summary>
+        #region Decrement
+        /// <summary>Decrements a specified variable and stores the result, as an atomic operation.</summary>
+        /// <param name="location">The variable whose value is to be decremented.</param>
+        /// <returns>The decremented value.</returns>
+        /// <exception cref="NullReferenceException">The address of location is a null pointer.</exception>
+        public static int Decrement(ref int location) =>
+            Add(ref location, -1);
+
+        /// <summary>Decrements a specified variable and stores the result, as an atomic operation.</summary>
+        /// <param name="location">The variable whose value is to be decremented.</param>
+        /// <returns>The decremented value.</returns>
+        /// <exception cref="NullReferenceException">The address of location is a null pointer.</exception>
+        [CLSCompliant(false)]
+        public static uint Decrement(ref uint location) =>
+            (uint)Add(ref Unsafe.As<uint, int>(ref location), -1);
+
+        /// <summary>Decrements a specified variable and stores the result, as an atomic operation.</summary>
+        /// <param name="location">The variable whose value is to be decremented.</param>
+        /// <returns>The decremented value.</returns>
+        /// <exception cref="NullReferenceException">The address of location is a null pointer.</exception>
+        public static long Decrement(ref long location) =>
+            Add(ref location, -1);
+
+        /// <summary>Decrements a specified variable and stores the result, as an atomic operation.</summary>
+        /// <param name="location">The variable whose value is to be decremented.</param>
+        /// <returns>The decremented value.</returns>
+        /// <exception cref="NullReferenceException">The address of location is a null pointer.</exception>
+        [CLSCompliant(false)]
+        public static ulong Decrement(ref ulong location) =>
+            (ulong)Add(ref Unsafe.As<ulong, long>(ref location), -1);
```

Comment:
Inlining decision is based on IL size right? This method is likely small enough, but will Unsafe.As affect downlevel inlining?